pulp_smart_proxy/app/views.py

```diff
@@ -0,0 +1,69 @@
+from urllib.parse import urljoin
+
+from django.conf import settings
+from drf_spectacular.utils import extend_schema
+from rest_framework.response import Response
+from rest_framework.views import APIView
+
+from pulpcore.app.apps import pulp_plugin_configs, get_plugin_config
+
+
+class FeaturesView(APIView):
+    """
+    Returns features of the smart_proxy
+    """
+
+    @extend_schema(
+        summary="Inspect features",
+        operation_id="features_read",
+    )
+    def get(self, request):
+        data = ["pulpcore"]
+        return Response(data)
+
+
+class FeaturesV2View(APIView):
+    """
+    Returns features of the smart_proxy in v2 format
+    """
+
+    @extend_schema(
+        summary="Inspect features",
+        operation_id="featuresv2_read",
+    )
+    def get(self, request):
+        host = settings.CONTENT_ORIGIN or request.get_host()
+        capabilities = [app.label for app in pulp_plugin_configs()]
+        data = {
+            "pulpcore": {
+                "http_enabled": False,
+                "https_enabled": True,
+                "settings": {
+                    "pulp_url": host,
+                    "mirror": settings.SMART_PROXY_MIRROR,
+                    "content_app_url": urljoin(host, settings.CONTENT_PATH_PREFIX),
+                    "username": settings.SMART_PROXY_AUTH_USERNAME,
+                    "password": settings.SMART_PROXY_AUTH_PASSWORD,
+                    "client_authentication": settings.SMART_PROXY_AUTH_METHODS,
```

Comment:
I was wondering if we could derive these values from settings. We set some values (https://github.com/theforeman/puppet-pulpcore/blob/8288f3736a1543417eba6322e7013d0f230bd85e/templates/settings.py.erb#L52-L59) so perhaps that works.

I think you should check if `settings.AUTHENTICATION_BACKENDS` and derive values based on that.

Reply:
I thought about that, but then wasn't sure how to get from `pulpcore.app.authentication.PulpNoCreateRemoteUserBackend` to `client_certificate` -- does the one imply the other?

Looking at https://pulpproject.org/pulpcore/docs/admin/guides/auth/external/, `PulpNoCreateRemoteUserBackend` does imply external auth, but the kind of external is up to the user. (OTOH, we're in the Foreman-domain here, and *we* probably can imply cert?)

Reply:
The non-cert option is mostly interesting for development setups, but I think it's highly questionable if it makes sense in this plugin.

In the Ruby based implementation we assume a trusted (authenticated) channel from Foreman to Foreman Proxy. The v2 features API (which exposes the settings) is normally only available if you are authenticated (typically using client certs).

If you take that to this, you either make the username/password available without authentication (effectively disabling authentication) or you have authentication but then don't need the credentials.

What you can do is to return `[]` if `pulpcore.app.authentication.PulpNoCreateRemoteUserBackend` isn't in there, but realistically the current implementation is probably OK.

Reply:
Is this related to https://github.com/pulp/pulpcore/pull/5438 which we could not bring forward until openapi 3.1.

Reply:
Sorta, but also not really :)

We have a Pulp running somewhere, and we need to tell the "client" (Katello) how to get there and how to auth there.
Today we do this by talking to the smart proxy and that returns a mostly static setting hash (pulp_url, client_authentication are the most important here) that gets used in other parts of the app.

So yeah, if the APIdoc would announce "you can use certs here", that would get rid of *that particular* line, the whole concept of "someone tells katello how pulp is configured" remains (I'd argue the way we do discovery of services, like pulp, is not optimal)